### PR TITLE
UPSTREAM: <carry>: cluster-api switch to using informers for cluster state

### PR DESCRIPTION
**Pull request by frobware**
This change moves away from iterating over all namespaces, then machinesets, then machines in each namespace (every 30s) to using a shared informer to provide the same information. This simplifies the code and it now looks very similar to many other controller/informer patterns.

### CODE REVIEW

**Comment by frobware** [comment]
/cc @derekwaynecarr @vikaschoudhary16 @bison @ingvagabund 

There will be follow-up changes once https://github.com/kubernetes-sigs/cluster-api/pull/565 merges. I didn't want to do that (or wait any longer) in this PR because the additional vendoring changes will be much larger compared to just adding the missing informers/listers generated packages as I have done here.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go [hunk 1]
```diff
@@ -62,79 +59,28 @@ func (ng *nodegroup) MaxSize() int {
...
-		return 0
-	}
-	glog.Infof("machineset: %q has %d replicas", ng.MachineSet.Name, *ng.MachineSet.Spec.Replicas)
-	return int(*ng.MachineSet.Spec.Replicas)
+	return int(ng.replicas)
 }
 
 func (ng *nodegroup) SetSize(nreplicas int) error {
-	ms, err := ng.clusterapi.MachineSets(ng.MachineSet.Namespace).Get(ng.MachineSet.Name, v1.GetOptions{})
+	machineSet, err := ng.provider.clusterapi.MachineSets(ng.namespace).Get(ng.name, v1.GetOptions{})
```
Comment (ingvagabund) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go]: You should not access method of indirect field. It makes components more coupled. `ng.provider.clusterapiClient().MachineSets()` is better.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/*
```
[the remark this answers is not in the document]
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go]: Ordinarily I agree. I was just falling in line with the existing pattern for the other cloud provider implementations. I guess they do it this way to make the logging obvious w.r.t. what cloud provider was running if/when issues are raised.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go [hunk 3]
```diff
@@ -18,41 +18,113 @@ package clusterapi
...
 
 const (
-	// ProviderName is the cloud nodegroup name for the cluster-api
-	// nodegroup.
-	ProviderName = "cluster-api"
-
-	refreshInterval = 30 * time.Second
+	// ProviderName is the name of cluster-api cloud provider.
+	ProviderName                  = "cluster-api"
+	nodeGroupMinSizeAnnotationKey = "sigs.k8s.io/cluster-api-autoscaler-node-group-min-size"
```
[the remark this answers is not in the document]
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go [hunk 4]
```diff
@@ -64,35 +136,63 @@ func (p *provider) GetResourceLimiter() (*cloudprovider.ResourceLimiter, error)
...
+
+	var result []cloudprovider.NodeGroup
 
-	for _, ms := range p.getClusterState().MachineSetMap {
-		if ms.MaxSize()-ms.MinSize() > 0 {
-			result = append(result, ms)
+	for _, ng := range nodegroups {
+		if ng.MaxSize()-ng.MinSize() > 0 {
+			glog.V(4).Infof("discovered machineset %q (min: %v, max: %v, replicas: %v)", ng, ng.minSize, ng.maxSize, ng.replicas)
+			result = append(result, ng)
```
[the remark this answers is not in the document]
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+	glog.V(4).Infof("waiting for machineset cache to sync")
+	if !cache.WaitForCacheSync(stopCh, c.machineSetInformer.Informer().HasSynced) {
+		return fmt.Errorf("cannot sync machineset cache")
+	}
+
+	return nil
+}
+
+func newClusterController(factory informers.SharedInformerFactory) (*clusterController, error) {
```
[the remark this answers is not in the document]
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go]: Done.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go [hunk 6]
```diff
@@ -18,41 +18,113 @@ package clusterapi
...
-	clusterSnapshotMutex sync.Mutex
-	clusterSnapshot      *clusterSnapshot
+	*clusterController
 
 	providerName    string
 	resourceLimiter *cloudprovider.ResourceLimiter
-	lastRefresh     time.Time
-	clusterapi      v1alpha1apis.ClusterV1alpha1Interface
-	kubeclient      *kubeclient.Clientset
+	clusterapi      clusterv1alpha1.ClusterV1alpha1Interface
```
[the remark this answers is not in the document]
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go (hunk 1 above)
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go]: This is all the same package. There's intrinsic coupling.

**Comment by frobware** [comment]
/test unit 

See also: https://github.com/kubernetes/autoscaler/pull/1456

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go (hunk 1 above)
Comment (frobware) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go]: Done (via dependency injection).

**Comment by ingvagabund** [comment]
Looks better, thanks for the changes.

**Pull request merged**